src/shogun/base/SGObject.h

```diff
@@ -1044,7 +1085,14 @@ class CSGObject
 	 */
 	template <class T>
 	void observe(const int64_t step, const std::string& name, const T& value,
-						  const AnyParameterProperties properties) const;
+						  const AnyParameterProperties properties) const
+	{
+		// If there are no observers attached, do not create/emit anything.
+		if (get_num_subscriptions() == 0) return;
+
+		auto obs = new ObservedValueTemplated<T>(step, name, static_cast<T>(clone_utils::clone(value)), properties);
```

Comment:
First off, very cool :"D
I have a question, how does this work? I thought that you can't call the constructor of a class that wasn't included.

Reply:
It is using a forward declaration (https://en.wikipedia.org/wiki/Forward_declaration). Basically, I can instantiate a pointer to that object here in the `.h` file and everything work. However, when I'll need to use the `observe()` function (inside a `.cpp` file), I will have to include the `ObservedValueTemplated.h` file to let the compiler know how that class is structured.

Reply:
Sorry this is not related to the PR, I'm just curious since I had a similar problem. 
But I believe a forward declared class is an incomplete type, and it shouldn't be possible to use any function of it including the constructor (it's obviously possible, hence the code compiles correctly) unless the header is included in the same place it is used. Like in this link: https://stackoverflow.com/questions/35263459/class-has-no-constructor-forward-declaration

Reply:
This is exactly what happens :) e.g. if you look at `SGObject.h`, there are some functions which uses `new ObservedValueTemplated()`. However, in each `.cpp` file in which those functions are used, you can see that there we include both the `SGObject.h` header (it can be included indirectly sometimes) and the `ObservedValueTemplated.h` header. Therefore, the compiler will have both the method and the class definition, and thus it will compile correctly.

Reply:
Oh, right, that makes sense. Thanks :")